### PR TITLE
Add useful methods to TrilinosWrappers::PreconditionAMG::AdditionalData

### PR DESCRIPTION
**Pull request by jppelteret**
These include:
- the creation of a parameter list, which can be later modified by the
user, and
- the initialisation of the null-space setting in an existing parameter
list.

Refers to the question asked in this forum post:
https://groups.google.com/forum/#!topic/dealii/rr6yvounKgs

### CODE REVIEW

**Comment by jppelteret** [comment]
What I've done here is to move the initialisation of the `Teuchos::ParameterList` out of the `PreconditionAMG::initialize()` function and into its own function in `PreconditionAMG::AdditionalData`. 

I've also split off the part that sets the null space of the operator. I thought that this might be useful in the case that the user wants to completely configure the parameter list by themselves, but wants to use a helper function set the null space (which seems to be quite a challenging thing to do correctly). Maybe this should be a separate function outside of `PreconditionAMG::AdditionalData`?

**Comment by jppelteret** [comment]
While I was busy here, I took the liberty of adding another constructor to `PreconditionAMG::AdditionalData` that sets the constant modes for a given `DoFHandler` and `ComponentMask`.

**Comment by jppelteret** [comment]
Hmm... this has unexpectedly lead to some test failures. Am investigating...
```
$ ctest -R precondition_amg_dgp -j4
Test project /home/jp/Programs/build/dealii
    Start 237: trilinos/precondition_amg_dgp_01.debug
    Start 238: trilinos/precondition_amg_dgp_01.release
    Start 239: trilinos/precondition_amg_dgp_02.debug
    Start 240: trilinos/precondition_amg_dgp_02.release
1/4 Test #240: trilinos/precondition_amg_dgp_02.release ...***Failed    0.81 sec
2/4 Test #238: trilinos/precondition_amg_dgp_01.release ...***Failed    0.83 sec
3/4 Test #239: trilinos/precondition_amg_dgp_02.debug .....***Failed    1.52 sec
4/4 Test #237: trilinos/precondition_amg_dgp_01.debug .....***Failed    1.56 sec

0% tests passed, 4 tests failed out of 4

Total Test time (real) =   1.62 sec

The following tests FAILED:
	237 - trilinos/precondition_amg_dgp_01.debug (Failed)
	238 - trilinos/precondition_amg_dgp_01.release (Failed)
	239 - trilinos/precondition_amg_dgp_02.debug (Failed)
	240 - trilinos/precondition_amg_dgp_02.release (Failed)
Errors while running CTest
```

```
$ ctest -R precondition_amg_dgp_01.debug -V
...
test 237
    Start 237: trilinos/precondition_amg_dgp_01.debug
...
237: ===============================   OUTPUT BEGIN  ===============================
...
237: Floating point exception (core dumped)
237: 
237: 
237: trilinos/precondition_amg_dgp_01.debug: ******    RUN failed    *******
```

**Comment by jppelteret** [comment]
Something for me to think about this evening... 

Set flag to print ML output:
```c++
  TrilinosWrappers::PreconditionAMG                 preconditioner;
  TrilinosWrappers::PreconditionAMG::AdditionalData data;
  DoFTools::extract_constant_modes(dof_handler,
                                   std::vector<bool>(1, true),
                                   data.constant_modes);
  data.smoother_sweeps = 2;
  data.output_details = true;
```
Current implementation (broken):
```
237: USing ParMETIS 3.x
237: Using `cg' for eigen-computations
237: *** 
237: *** ML_Epetra::MultiLevelPreconditioner
237: ***
237: Matrix has 12288 rows and 182016 nonzeros, distributed over 1 process(es)
237: The linear system matrix is an Epetra_CrsMatrix
237: ** Leaving column map of Main linear system matrix unchanged
237: Default values for `SA'
237: Maximum number of levels = 10
237: Using increasing levels. Finest level  = 0, coarsest level = 9
237: Number of applications of the ML cycle = 1
237: Number of PDE equations = 1
237: Aggregation threshold = 0.0001
237: Max coarse size = 2000
237: R and P smoothing : P = (I-\omega A) P_t, R = P^T
237: R and P smoothing : \omega = 1.333/lambda_max
237: Null space type      = user-supplied
237: Null space dimension = 1
237: -----------------------------------------------------------------------
237: ML_Gen_MultiLevelHierarchy (level 0) : Gen Restriction and Prolongator
237: -----------------------------------------------------------------------
237: ML_Aggregate_Coarsen (level 0) begins
237: ML_Aggregate_CoarsenUncoupled : current level = 0
237: ML_Aggregate_CoarsenUncoupled : current eps = 1.000000e-04
237: Aggregation(UVB) : Total nonzeros = 182016 (Nrows=12288)
237: Aggregation(UC) : Phase 0 - no. of bdry pts  = 0 
237: Aggregation(UC) : Phase 1 - nodes aggregated = 8573 (12288)
237: Aggregation(UC) : Phase 1 - total aggregates = 1510 
237: Aggregation(UC_Phase2_3) : Phase 1 - nodes aggregated = 8573 
237: Aggregation(UC_Phase2_3) : Phase 1 - total aggregates = 1510
237: Aggregation(UC_Phase2_3) : Phase 2a- additional aggregates = 0
237: Aggregation(UC_Phase2_3) : Phase 2 - total aggregates = 1510
237: Aggregation(UC_Phase2_3) : Phase 2 - boundary nodes   = 0
237: Aggregation(UC_Phase2_3) : Phase 3 - leftovers = 0 and singletons = 0
237: Floating point exception (core dumped)
```
Some correct output:
```
$ cat tests/trilinos/precondition_amg_dgp.debug/stdout 
------------------------------------------------------------------------------
USing ParMETIS 3.x
Using `cg' for eigen-computations
*** 
*** ML_Epetra::MultiLevelPreconditioner
***
Matrix has 12288 rows and 182016 nonzeros, distributed over 1 process(es)
The linear system matrix is an Epetra_CrsMatrix
** Leaving column map of Main linear system matrix unchanged
Default values for `SA'
Maximum number of levels = 10
Using increasing levels. Finest level  = 0, coarsest level = 9
Number of applications of the ML cycle = 1
Number of PDE equations = 1
Aggregation threshold = 0.0001
Max coarse size = 2000
R and P smoothing : P = (I-\omega A) P_t, R = P^T
R and P smoothing : \omega = 1.333/lambda_max
Null space type      = user-supplied
Null space dimension = 1
-----------------------------------------------------------------------
ML_Gen_MultiLevelHierarchy (level 0) : Gen Restriction and Prolongator
-----------------------------------------------------------------------
ML_Aggregate_Coarsen (level 0) begins
ML_Aggregate_CoarsenUncoupled : current level = 0
ML_Aggregate_CoarsenUncoupled : current eps = 1.000000e-04
Aggregation(UVB) : Total nonzeros = 182016 (Nrows=12288)
Aggregation(UC) : Phase 0 - no. of bdry pts  = 0 
Aggregation(UC) : Phase 1 - nodes aggregated = 8573 (12288)
Aggregation(UC) : Phase 1 - total aggregates = 1510 
Aggregation(UC_Phase2_3) : Phase 1 - nodes aggregated = 8573 
Aggregation(UC_Phase2_3) : Phase 1 - total aggregates = 1510
Aggregation(UC_Phase2_3) : Phase 2a- additional aggregates = 0
Aggregation(UC_Phase2_3) : Phase 2 - total aggregates = 1510
Aggregation(UC_Phase2_3) : Phase 2 - boundary nodes   = 0
Aggregation(UC_Phase2_3) : Phase 3 - leftovers = 0 and singletons = 0
Aggregation(UC) : QR factorization - too small aggregates = 0
Calculating eigenvalue estimate using CG method
Gen_Prolongator (level 0) : Max eigenvalue = 1.9700e+00

Prolongator/Restriction smoother (level 0) : damping factor #1 = 6.7663e-01
Prolongator/Restriction smoother (level 0) : ( = 1.3330e+00 / 1.9700e+00)

Smoothed Aggregation : operator complexity = 1.160272e+00.
Time to build the hierarchy = 0.0137411 (s)
Number of actual levels : 2

Smoother (level 0) : # global rows = 12288, # estim. global nnz = 182016, # nnz per row = 14.81
Smoother (level 0) : MLS/Chebyshev, polynomial order = 2, alpha = 10, both
Smoother (level 0) : lambda_min = 0.00227521, lambda_max = 1.97004
Smoother (level 0) : Setup time : 5.9443e-05 (s)

Amesos (level 1) : NumGlobalRows = 1510
Amesos (level 1) : NumGlobalNonzeros = 29172
Amesos (level 1) : Fill-in = 1.27942 %
Amesos (level 1) : Building KLU
Amesos (level 1) : Time for factorization  = 0.00803115 (s)


Cumulative timing for construction so far: 
- for initial setup   = 0.0194162 (s)
- for hierarchy setup = 0.0137747 (s)
- for smoothers setup = 5.9443e-05 (s)
- for coarse setup    = 0.00833387 (s)
- for final setup     = 3.3938e-05 (s)
Total for this setup  = 0.0417394 (s)
------------------------------------------------------------------------------
------------------------------------------------------------------------------
...

```


This has me a bit stumped, because I really believe that I have mostly just moved some code around. I don't see how this is any different to the original implementation :-/

**Comment by jppelteret** [comment]
I worked out what was going wrong -- this `distributed_constant_modes` vector needed to remain in scope until initialise() was called. So I've added an argument that stores the initialised vector, which the user will have to ensure remains in scope for the necessary duration.

**Comment by jppelteret** [comment]
@kronbichler Would you mind please reviewing this? I mostly moved existing code around and duplicated a test, so I don't think that it will take much effort. I'll point out the biggest changes I made to the code.

**Comment by kronbichler** [comment]
@jppelteret I will look into this in the coming days, but probably not today.

**Comment by jppelteret** [comment]
Thanks a lot!

File: source/lac/trilinos_precondition_ml.cc [hunk 1]
```diff
@@ -190,13 +215,75 @@ namespace TrilinosWrappers
...
           parameter_list.set("null space: vectors",
                              distributed_constant_modes.Values());
-        // We need to set a valid pointer to data even if there is no data on
-        // the current processor. Therefore, pass a dummy in that case
         else
-          parameter_list.set("null space: vectors", dummy.data());
+          {
+            // We need to set a valid pointer to data even if there is no data
+            // on the current processor. Therefore, pass a dummy in that case
+            static std::vector<double> dummy;
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply source/lac/trilinos_precondition_ml.cc]: I would prefer that variant.

File: include/deal.II/lac/trilinos_precondition.h [hunk 2]
```diff
@@ -1382,6 +1385,101 @@ namespace TrilinosWrappers
...
+       *
+       * This constructor initializes the @p constant_modes for the @p dof_handler
+       * and the components chosen by @p component_mask . See the documentation
+       * of the @p constant_modes parameter and
+       * DoFTools::extract_constant_modes() function for more information on
+       * this point.
+       */
+      template <typename DoFHandlerType>
+      AdditionalData(const DoFHandlerType &dof_handler,
+                     const ComponentMask & component_mask,
```
Comment (kronbichler) [inline include/deal.II/lac/trilinos_precondition.h]: I am not sure how we would like to handle this case, but it feels that this change goes somewhat against our directory philosophy of not including any FE-specific stuff in a linear algebra class. We probably won't ever do that, but if we wanted to break out the `base/` and `lac/` directories we could still manage that (we only have a handful of dependencies in `base`), but every change like this one would make it more difficult. How are the opinions on this?

File: source/lac/trilinos_precondition_ml.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jppelteret) [reply source/lac/trilinos_precondition_ml.cc]: I've removed the dummy vector now. As far as I understand, `distributed_constant_modes.Values()` should always point to something valid. Or is this not correct?

File: include/deal.II/lac/trilinos_precondition.h (hunk 2 above)
Comment (jppelteret) [reply include/deal.II/lac/trilinos_precondition.h]: I wasn't conscious of the fact that we tried to keep these separate.  I can just drop this commit to keep things moving...

**Comment by jppelteret** [comment]
@kronbichler I dropped the contentious commit, so I guess this is ready for you to look at again when you have some time to do so.

**Pull request merged**